### PR TITLE
add Kuadrant's AuthPolicy

### PR DESCRIPTION
**Pull request by rahulanand16nov**
#133 
```yaml
apiVersion: apim.kuadrant.io/v1alpha1
kind: AuthPolicy
metadata:
  name: toystore
spec:
  targetRef:
    group: gateway.networking.k8s.io
    kind: HTTPRoute
    name: toystore
  rules:
  - to:
    - operation:
        methods: ["GET"]
        paths: ["/toy"]
  action: CUSTOM
  provider:
    name: kuadrant-authorization
```

-----
## Verification Steps:
1. `make local-setup`
2. `kubectl apply -f examples/toystore/toystore.yaml`
3. `kubectl apply -f examples/toystore/httproute.yaml`
4. Configure Authorino
```
kubectl apply -f -<<EOF
apiVersion: authorino.kuadrant.io/v1beta1
kind: AuthConfig
metadata:
  name: toystore-protection
spec:
  hosts:
  - api.toystore.com
  identity:
  - name: friends
    apiKey:
      labelSelectors:
        group: friends
    credentials:
      in: authorization_header
      keySelector: APIKEY
EOF

kubectl apply -f -<<EOF
apiVersion: v1
kind: Secret
metadata:
  name: api-kek
  labels:
    authorino.kuadrant.io/managed-by: authorino
    group: friends
stringData:
  api_key: ITSFORDEMO
type: Opaque
EOF
```
5. `kubectl apply -f config/samples/apim_v1alpha1_authpolicy.yaml`
```
kubectl get authorizationpolicy -n kuadrant-system
NAME                                          AGE
on-kuadrant-gwapi-gateway-using-hr-toystore   3s
```
7. `curl -v -H 'Host: api.toystore.com' http://localhost:9080/toy` // unauthorized
8. `curl -v -H 'Authorization: APIKEY ITSFORDEMO' -H 'Host: api.toystore.com' http://localhost:9080/toy` // sweet 200 OK :)
9. `kubectl delete -f config/samples/apim_v1alpha1_authpolicy.yaml`
```
kubectl get authorizationpolicy -n kuadrant-system
No resources found in kuadrant-system namespace.
```

## Next Step

- [ ] Target Gateway resource.
- [ ] Take hosts from GWAPI resources and limit its configurability.

### CODE REVIEW

File: controllers/apim/authpolicy_controller.go [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+	return ctrl.Result{}, nil
+}
+
+// IstioAuthPolicy generates Istio's AuthorizationPolicy using Kuadrant's AuthPolicy
+func (r *AuthPolicyReconciler) reconcileAuthPolicy(ctx context.Context, ap *apimv1alpha1.AuthPolicy) error {
+	logger := logr.FromContext(ctx)
+
+	httpRoute, err := r.fetchHTTPRoute(ctx, ap)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
```
Comment (eguzki) [inline controllers/apim/authpolicy_controller.go:91]: if HTTPRoute is not found, shouldn't the controller at least retry?

**Review by eguzki — APPROVED** [review]

File: controllers/apim/authpolicy_controller.go (hunk 1 above)
Comment (rahulanand16nov) [reply controllers/apim/authpolicy_controller.go:91]: There is a possibility of a loop right then? when HTTPRoute won't exist even in the future.

File: controllers/apim/authpolicy_controller.go [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+
+	logger.Info("removed Istio's AuthorizationPolicy")
+	return nil
+}
+
+// fetchHTTPRoute fetches the HTTPRoute described in targetRef *within* AuthPolicy's namespace.
+func (r *AuthPolicyReconciler) fetchHTTPRoute(ctx context.Context, ap *apimv1alpha1.AuthPolicy) (*gatewayapiv1alpha2.HTTPRoute, error) {
+	logger := logr.FromContext(ctx)
+	key := client.ObjectKey{
+		Name:      string(ap.Spec.TargetRef.Name),
```
Comment (markmc) [inline controllers/apim/authpolicy_controller.go:185]: I don't see any check that this TargetRef is actually to a HTTPRoute? Am I missing something? HTH

Comment (rahulanand16nov) [reply controllers/apim/authpolicy_controller.go:185]: Hey Mark, It's still in the development stages, but right now, we allow (assume) it to only target an HTTPRoute, but with the next PR, it should be able to target both `Gateway` and `HTTPRoute`.

Comment (rahulanand16nov) [reply controllers/apim/authpolicy_controller.go:185]: You can find the validation checks in #146

Comment (markmc) [reply controllers/apim/authpolicy_controller.go:185]: Thanks, I just found the Validate() methods while looking at the RateLimitPolicy example!

**Pull request merged**